samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala

```diff
@@ -801,21 +804,40 @@ class SamzaContainer(
 
   def addShutdownHook {
     val runLoopThread = Thread.currentThread()
-    Runtime.getRuntime().addShutdownHook(new Thread() {
+    shutdownHookThread = new Thread("CONTAINER-SHUTDOWN-HOOK") {
       override def run() = {
         info("Shutting down, will wait up to %s ms" format shutdownMs)
         runLoop match {
           case runLoop: RunLoop => runLoop.shutdown
           case asyncRunLoop: AsyncRunLoop => asyncRunLoop.shutdown()
         }
-        runLoopThread.join(shutdownMs)
-        if (runLoopThread.isAlive) {
-          warn("Did not shut down within %s ms, exiting" format shutdownMs)
-        } else {
-          info("Shutdown complete")
+        try {
+          runLoopThread.join(shutdownMs)
+        } catch {
+          case e: Exception => // Ignore to avoid deadlock with uncaughtExceptioHandler. See SAMZA-1220
+        } finally {
+          if (runLoopThread.isAlive) {
+            warn("Did not shut down within %s ms, exiting" format shutdownMs)
+          } else {
+            info("Shutdown complete")
+          }
         }
       }
-    })
+    }
+    Runtime.getRuntime().addShutdownHook(shutdownHookThread)
+  }
+
+  def removeShutdownHook = {
+    try {
+      if (shutdownHookThread != null) {
+        Runtime.getRuntime.removeShutdownHook(shutdownHookThread)
+      }
+    } catch {
+      case e: IllegalStateException => {
+        // When samza is shutdown by external command, IllegalStateException will be thrown.
```

Comment:
Maybe rephrase to:
"Thrown when then JVM is already shutting down, so safe to ignore."?